proofs/src/circuit/groups.rs

```diff
@@ -0,0 +1,1143 @@
+//! Traits and structs related to groups of regions.
```

Comment:
This module has all the support types and traits for the group API. More than half of it are unit tests so is less code than it seems at first. 

Also, this file alone accounts for a third of the PR